Add DB table structure for custom order tables.

vedanshujain opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This adds the base table structure for the custom order table as described in [the developer blog post](https://developer.woocommerce.com/2022/01/17/the-plan-for-the-woocommerce-custom-order-table/). The tool in _Settings > Tools_ has now two functions: "Create the tables" and "Delete the tables" (so no more references to migration, syncrhonization, etc; these will be handled by/displayed in settings once implemented).

Additionally, a new `DatabaseUtil` class is added which contains (amongst other stuff) a refactored version of the `verify_db` tool.

Closes #31612.

### How to test the changes in this Pull Request:

1. Enable the custom orders tables feature by adding the code snippet: 

```php
wc_get_container()
  ->get('Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController')
  ->show_feature();
```

2. Go to WooCommerce > Settings > Advanced, you'll see a new "Custom data stores" section with a message stating that there's nothing to configure yet:

![image](https://user-images.githubusercontent.com/937723/153186154-ca1f8fd6-29a8-48c4-8a18-7c4d8f44d249.png)

2. In a separate browser tab go to WooCommerce > Status > Tools, find a tool named "Create the custom orders tables" and run it, it will then become a "Delete the custom orders tables" tool.

Before running:

![image](https://user-images.githubusercontent.com/937723/153186375-6e1545ec-903a-41b5-ad55-a86688670a9f.png)

After running:

![image](https://user-images.githubusercontent.com/937723/153188293-739e2234-be4a-4183-a5c7-e9c1d4c231e6.png)

![image](https://user-images.githubusercontent.com/937723/153186885-10ee0af6-12cb-40f9-9443-87ebaa4d35b1.png)

3. Verify that the database tables have been created (`wp_wc_orders`, `wp_wc_order_addresses` and `wp_wc_order_operational_data`).

4. Go back to the settings tab, reload and you'll now see a setting that allows you to enable the table usage. Enable it and save.

![image](https://user-images.githubusercontent.com/937723/153365717-77260ae1-bcef-4040-84aa-7179d7a219ef.png)

5. Go back to the tools page, DON'T reload it and click "Save". You should get an error and you'll see that the "Delete" button is now disabled.

![image](https://user-images.githubusercontent.com/937723/153188759-72c8e528-332c-483b-aa11-06aaf9ba01fe.png)

![image](https://user-images.githubusercontent.com/937723/153188906-c8672e28-12e9-43c3-a9d6-d2a8f8a3c49a.png)

6. Go back to settings, disable the table usage and save.

7. Go back to the tools page, reload it, the "Delete" button should now be enabled. Run the tool.

8. Verify that the tables have been indeed deleted, and that reloading the settings page shows the "Create the tables firsts..." message again.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Enhancement - Add DB table structure for custom order tables.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

vedanshujain committed:
Refactor verify_base_tables method to allow custom schema verification.

vedanshujain committed:
Add initial DB structure for custom order tables.

vedanshujain committed:
Fixups for typo and unintended code.

Konamiman commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -368,6 +361,27 @@ public static function verify_base_tables( $modify_notice = true, $execute = fal
+	 *
+	 * return array List of missing tables.
+	 */
+	public static function verify_schema( $schema ) {
```
If we are going to add this as a public method I think we need to come up with a better name, `verify_schema` is too generic and the same for the argument. Maybe something like `verify_database_tables_exist( $tables_creation_sql )`?

Or even better, since this method is public solely because we need to invoke it from another class that is `Internal`, why don't we move this method to an internal class too (maybe something like `DbUtils`), then invoke it from inside `verify_database_tables` too? That would require a bit of setup with the DI container though, but I can help with that.

vedanshujain replied:
Thanks for the suggestion. Moved this method to a new DBUtil class in https://github.com/woocommerce/woocommerce/pull/31811/commits/3d29e8329c1196e5dd5b40214df3b324bc9ed602. Although it's still public since we do want folks to use it.

Konamiman commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -65,10 +62,10 @@ public function data_regeneration_is_in_progress(): bool {
 		update_option( self::CUSTOM_ORDERS_TABLE_DATA_REGENERATION_DONE_COUNT, 0 );
-
-		// TODO: Create the tables as appropriate, schedule the table filling in batches with Action Scheduler.
+		dbDelta( $this->data_store->get_schema() );
```
This is another case where a separate `DbUtils` class would be appropriate I think. The `require_once` and the `dbDelta` could be properly encapsulated in a dedicated method in such a class, and this would also help later with unit tests.

vedanshujain replied:
Done in https://github.com/woocommerce/woocommerce/pull/31811/commits/3d29e8329c1196e5dd5b40214df3b324bc9ed602

Konamiman commented on plugins/woocommerce/src/Internal/DataStores/Orders/FeatureController.php:
```diff
@@ -271,7 +271,7 @@ private function delete_custom_orders_table() {
 	 * @return array The updated settings sections array.
 	 */
 	private function get_settings_sections( array $sections ): array {
-		if ( ! $this->is_feature_visible() || ! $this->data_synchronizer->check_orders_table_exists() ) {
```
The expected user flow at this early stage of development is:

1. User adds the code snippet to enable the feature.
2. User uses the dedicated tool to create the table.
3. User can then go and see and modify the related settings.

So, 3 doesn't make much sense until 1 and 2 have been done. Given that, I think that the `check_orders_table_exists` check should remain for now (for sure we'll remove it in the future, but not yet).

vedanshujain replied:
As discussed, I have changed a bit to still show the setting if feature is enabled, but also added a message to go to the tools to generate tables when they don't exists in https://github.com/woocommerce/woocommerce/pull/31811/commits/fc0c45a204aad1a2d629351ca40acca11187e712.

Konamiman commented on plugins/woocommerce/src/Internal/DataStores/Orders/FeatureController.php:
```diff
@@ -289,7 +289,7 @@ private function get_settings_sections( array $sections ): array {
 	 * @return array The updated settings array.
 	 */
 	private function get_settings( array $settings, string $section_id ): array {
-		if ( ! $this->is_feature_visible() || 'custom_data_stores' !== $section_id || ! $this->data_synchronizer->check_orders_table_exists() ) {
```
Same here, `check_orders_table_exists` should be left in place for now.

vedanshujain replied:
See https://github.com/woocommerce/woocommerce/pull/31811#discussion_r801827095

Konamiman commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -19,10 +21,19 @@ class OrdersTableDataStore extends \Abstract_WC_Order_Data_Store_CPT implements
+
 	// TODO: Add methods for other table names as appropriate.
 
 	//phpcs:disable Squiz.Commenting.FunctionComment.Missing
```
Is it ok to do this? I thought it was a big sin in our codebase 🙂

vedanshujain replied:
It's not ideal, but I am trying to figure out a good way for custom data stores to declare tables and schema. Bit of a WIP, so kept it as it is for now.

Konamiman committed:
Merge branch 'scafolding-for-custom-order-tables' into mvp-cot/intial-table-strucuture

vedanshujain committed:
Refactor and move DB related methods to util class.

vedanshujain committed:
Show link to create custom orders table.

Konamiman commented on plugins/woocommerce/src/Utilities/DBUtil.php:
```diff
@@ -0,0 +1,46 @@
+
+use Automattic\Jetpack\Constants;
+
+class DBUtil {
```
Given that the methods in this class are not idempotent (since they depend on the database state), I think it should have non-static methods and be instantiated via dependency injection container. This will ease the task of writing unit tests, as we'll be able to easily mock the database behavior.

Also I know I suggested the `DBUtil` name myself, but maybe a more explicit name like `DatabaseUtil` would be more appropriate.

Konamiman commented on plugins/woocommerce/src/Utilities/DBUtil.php:
```diff
@@ -0,0 +1,46 @@
+	 *
+	 * return array List of missing tables.
+	 */
+	public static function verify_database_tables_exist( $table_creation_sql ) {
```
What about naming this `get_missing_database_tables` instead? This way it's more clear what the returned value is.

Konamiman commented on plugins/woocommerce/src/Utilities/DBUtil.php:
```diff
@@ -0,0 +1,46 @@
+	 *
+	 * @return array List of tables that we were not able to create.
+	 */
+	public static function create_database_tables( $table_creation_sql ) {
```
Suggested refactoring:

* Make this method a direct wrapper of `dbDelta`, that is, make it accept a second argument to determine whether the SQL is actually executed or not; and return the value given by `dbDelta` directly. This will make this method helpful as a direct replacement of `dbDelta` elsewhere in the codebase, when that method is used and we want to add unit tests around it. (Also name it `db_delta` to follow our standards)

* Add an additional `parse_dbdelta_output` method that converts the output from `dbDelta` into a list of missing tables (or into a more complex data structure if that output really returns something else that's useful).

Konamiman committed:
Make the db utils class non-static, change the custom orders tables UI

- Rename DBUtils to DatabaseUtils, and register it in the DI container
  (instead of having it just contain static methods)

- The tool for the custom orders tables has now two shapes:
  "create tables" and "delete tables" (so no regeneration, migration...)

- Not yet used constants and methods from DataSynchronizer removed

- Added missing method comments

vedanshujain commented on plugins/woocommerce/src/Utilities/DatabaseUtil.php:
```diff
@@ -0,0 +1,72 @@
+		}
+
+		//phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+		$wpdb->query( "DROP TABLE IF EXISTS {$table_name}" );
```
we should return the output here, also we need to escape the table name

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -282,39 +247,41 @@ private function get_settings_sections( array $sections ): array {
-				'title'         => __( 'Enable table usage', 'woocommerce' ),
-				'desc'          => __( 'Use the custom orders table as the main orders data store.', 'woocommerce' ),
+				'title'         => __( 'Enable tables usage', 'woocommerce' ),
+				'desc'          => __( 'Use the custom orders tables as the main orders data store.', 'woocommerce' ),
```
During development I think we should add a warning here actually:
```suggestion
				'desc'          => __( 'WARNING: This feature is currently under development and will cause database instability. For contributors only.', 'woocommerce' ),
```

This would handle the case where a normal user would try to enable the feature on an old WooCommerce version.

Konamiman replied:
Ok, I have implemented it as an extra message on top of the setting instead of as a replacement of the original message.

vedanshujain commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -35,7 +35,7 @@ class CustomOrdersTableController {
 	 * @var DataSynchronizer
 	 */
-	private $data_synchronizer;
+	public $data_synchronizer;
```
unintended change sneaked in, probably from my code.

vedanshujain reviewed (commented):
The code looks good to me, added a few minor comments. I will test it out a bit before approving.

Konamiman committed:
Small changes to the custom orders table implementation

- Add an extra warning message in the settings page
- Escape table name for DROP statement in DatabaseUtil::drop_database_table

Konamiman committed:
Remove no longer needed 'use'.

dokker commented on plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableDataStore.php:
```diff
@@ -140,4 +154,74 @@ public function get_order_item_type( $order, $order_item_id ) {
+	tax_amount decimal(26,8) null,
+	total_amount decimal(26,8) null,
+	customer_id bigint(20) unsigned null,
+	billing_email varchar(320) null,
```
When somebody use INNODB utf8mb4 this value exceeds the limit. In this case the max accepted VARCHAR length is 191. Otherwise it fails with "Specified key was too long; max key length is 1000 bytes" error.

ilchybaby commented:
I get a strange error not sure what I'm doing wrong....

`[18-May-2022 00:38:48 UTC] PHP Fatal error:  Uncaught Error: Cannot access private const Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController::CUSTOM_ORDERS_TABLE_USAGE_ENABLED_OPTION in /home/smtoyw1/public_html/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:158
Stack trace:
#0 /home/smtoyw1/public_html/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php(129): Automattic\WooCommerce\Internal\DataStores\Orders\DataSynchronizer->get_current_orders_pending_sync_count()
#1 /home/smtoyw1/public_html/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php(307): Automattic\WooCommerce\Internal\DataStores\Orders\DataSynchronizer->get_sync_status()
#2 /home/smtoyw1/public_html/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php(90): Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController->get_settings(Array, 'custom_data_sto...')
#3 /home/smtoyw1/public_html/wp-includes/ in /home/smtoyw1/public_html/wp-content/plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php on line 158
`

Konamiman commented:
Hi @ilchybaby, are you still having the issue with the latest version of WooCommerce?
